dumpling/dumpling.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dumpling
+
+import (
+	"context"
+	"os"
+	"strings"
+	"time"
+
+	"github.com/pingcap/dm/dm/config"
+	"github.com/pingcap/dm/dm/pb"
+	"github.com/pingcap/dm/dm/unit"
+	"github.com/pingcap/dm/pkg/log"
+
+	"github.com/pingcap/dumpling/v4/export"
+	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
+	"github.com/siddontang/go/sync2"
+	"go.uber.org/zap"
+)
+
+// Dumpling dumps full data from a MySQL-compatible database
+type Dumpling struct {
+	cfg *config.SubTaskConfig
+
+	logger log.Logger
+
+	dumpConfig *export.Config
+	closed     sync2.AtomicBool
+}
+
+// NewDumpling creates a new Dumpling
+func NewDumpling(cfg *config.SubTaskConfig) *Dumpling {
+	m := &Dumpling{
+		cfg:    cfg,
+		logger: log.With(zap.String("task", cfg.Name), zap.String("unit", "dump")),
+	}
+	return m
+}
+
+// Init implements Unit.Init
+func (m *Dumpling) Init(ctx context.Context) error {
+	var err error
+	m.dumpConfig, err = m.constructArgs()
+	return err
+}
+
+// Process implements Unit.Process
+func (m *Dumpling) Process(ctx context.Context, pr chan pb.ProcessResult) {
+	dumplingExitWithErrorCounter.WithLabelValues(m.cfg.Name).Add(0)
+
+	failpoint.Inject("dumpUnitProcessWithError", func(val failpoint.Value) {
+		m.logger.Info("dump unit runs with injected error", zap.String("failpoint", "dumpUnitProcessWithError"), zap.Reflect("error", val))
+		msg, ok := val.(string)
+		if !ok {
+			msg = "unknown process error"
+		}
+		pr <- pb.ProcessResult{
+			IsCanceled: false,
+			Errors:     []*pb.ProcessError{unit.NewProcessError(errors.New(msg))},
+		}
+		failpoint.Return()
+	})
+
+	begin := time.Now()
+	errs := make([]*pb.ProcessError, 0, 1)
+	isCanceled := false
+
+	failpoint.Inject("dumpUnitProcessForever", func() {
+		m.logger.Info("dump unit runs forever", zap.String("failpoint", "dumpUnitProcessForever"))
+		<-ctx.Done()
+		failpoint.Return()
+	})
+
+	// NOTE: remove output dir before start dumping
+	// every time re-dump, loader should re-prepare
+	err := os.RemoveAll(m.cfg.Dir)
+	if err != nil {
+		m.logger.Error("fail to remove output directory", zap.String("directory", m.cfg.Dir), log.ShortError(err))
```

Comment:
do we need to append err to `errs` and then return? If the old dir is not removed, may load wrong data to upstream